### PR TITLE
Polygon tracking feature with XMem tracker

### LINKED ISSUE #5465: enhancement: XMem mask video propagation as CVAT AI Tools polygon tracker
### My actions before raising this issue
- I read/searched [the docs](https://github.com/cvat-ai/cvat/tree/master#documentation)
- I searched [past issues](/issues)

There is a great instance segmentation mask propagation tool `[ECCV 2022] XMem: Long-Term Video Object Segmentation with an Atkinson-Shiffrin Memory Model` by Deep learning researchers Yue Zhao and Rex Cheng. 

We have a video, and one instance segmentation mask for the first frame of our video.
We give them to the neural network and it predicts masks for all other frames.

There is a link to git repo [link](https://github.com/hkchengrex/XMem).

They have a Colab demo with their neural network [link](https://colab.research.google.com/drive/1RXK5QsUo2-CnOiy5AOSjoZggPVHOPh1m?usp=sharing ) There are easy launch and simple code that can be integrated.
The example from the link is great for creating a potential CVAT interactor of mask polygon tracker type )

There are some demo om video:
* [link1](https://user-images.githubusercontent.com/7107196/177921527-7a1bd593-2162-4598-9adf-f2112763fccf.mp4)
* [link2](https://user-images.githubusercontent.com/7107196/177920383-161f1da1-33f9-48b3-b8b2-09e450432e2b.mp4)

It works great and looks as fantastic.

This DL model also have a GUI. We can annotate mask for some objects and then propagate masks for them through all video.
It's interesting that the creators use f-BRS for creating instance segmentation mask for future mask fropagation.
CVAT have f-BRS as segmentation interactor too.
But in general, it is not so important what the preliminary mask is created with. 
I think other CVAT's AI tool's interactors will work great too( for example HRnet).

It seems like a very instruction tool for CVAT for video instance segmentation tool.

### Context
Get a powerful instance segmentation annotation tool for video )

### PR DESCRIPTION
**Pull request by omerferhatt**
### Motivation and context
Polygon tracking is a crucial part of computer vision annotation tasks. It helps the required consecutive frame annotation, shortens the required annotation time, creates more precise annotation and moreover simplifies all HITL (Human in the loop) processes. Since the CVAT doesn't support this feature, it seems that adding this functionality with side benefits will be good for the sake of this project.

It aims to solve the below problems:

- Using polygon tracking algorithms natively inside the CVAT to use it with semi-automatic annotations.
- Tracking multiple polygon objects without losing the common track options (enable/disable, modify the last tracked object and continue from it, etc.)
- Creating trackable polygon objects through `interactor`'s (which allows to use `SAM` with any kind of polygon tracking algorithm)

According to the previous issues, it:

- Fixes #5465 
- Fixes https://github.com/cvat-ai/cvat/issues/5984#issuecomment-1509587246
- Fixes https://github.com/cvat-ai/cvat/issues/5984#issuecomment-1510716028
- Helps #4915 
- Helps #5686

#### How it looks
[Initial demo video](https://github.com/cvat-ai/cvat/assets/49915636/6b6cb602-a622-4d73-a6d5-521f110d7ea4)

- SAM runs on the CPU and XMem tracker runs on the GPU

#### Resource usage
- For 5 object tracking on 1280x720 video will use approx. 2GB VRAM at peak. 

#### Known issues and further optimization cases:

- When `maxWorker` is more than 1 in `function.yaml`, the entire tracker works again in a duplicate manner. Since the problem is not solved, it is set to 1 for now.
- Since tracking works on individual objects, it can be improved with batch inference.
- As explained in step 2, this increases the request size since model states are kept separately for each input. Therefore, the process outside the model slows down as the number of objects tracked increases.

### How has this been tested?
Tested and developed according to [CVAT documentation](https://docs.cvat.ai/docs/contributing/development-environment/)

### Checklist

- [ ] I submit my changes into the `develop` branch
- [ ] I have created a changelog fragment 
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [x] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary
  ([cvat-canvas](https://github.com/cvat-ai/cvat/tree/develop/cvat-canvas#versioning),
  [cvat-core](https://github.com/cvat-ai/cvat/tree/develop/cvat-core#versioning),
  [cvat-data](https://github.com/cvat-ai/cvat/tree/develop/cvat-data#versioning) and
  [cvat-ui](https://github.com/cvat-ai/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/cvat-ai/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

## Summary by CodeRabbit

- **New Features**
  - Added object type selection functionality to enhance user control over shape and polygon annotations.
  - Introduced new serverless functions for advanced object segmentation using a PyTorch-based model with GPU support.

- **Enhancements**
  - Improved filtering logic to support both rectangles and polygons more effectively in the annotation interface.

- **Documentation**
  - Updated component documentation to reflect new functionalities and logic adjustments.

- **Performance**
  - Optimized image processing and object tracking within the serverless PyTorch model environment.

### CODE REVIEW

**Comment by PMazarovich** [comment]
Hello, @bsekachev , do you think this MR will be reviewed/merged?
This sounds very helpful for annotators
@omerferhatt thanks!

**Comment by bsekachev** [comment]
I will try to review it on this week.

File: cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/tools-control.tsx [hunk 1]
```diff
@@ -1358,6 +1387,7 @@ export class ToolsControlComponent extends React.PureComponent<Props, State> {
                 <Tabs type='card' tabBarGutter={8}>
                     <Tabs.TabPane key='interactors' tab='Interactors'>
                         {this.renderMasksConvertingBlock()}
+                        {convertMasksToPolygons ? this.renderObjectTypeBlock() : null}
```
Comment (bsekachev) [inline cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/tools-control.tsx:1390]: `xmem` is a tracker, not interactor. The difference between them is exactly that interactors produce shapes on one frame, when tracker produces shapes on multiple frames (tracks). 
Why are you modifying the block responsible for interactors?

File: cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/tools-control.tsx [hunk 2]
```diff
@@ -958,6 +963,29 @@ export class ToolsControlComponent extends React.PureComponent<Props, State> {
         );
     }
 
+    private renderObjectTypeBlock(): JSX.Element {
+        const { selectedObjectType } = this.state;
+        const objectTypes = Object.values(ObjectType);
+        objectTypes.splice(objectTypes.indexOf(ObjectType.TAG), 1);
```
Comment (bsekachev) [inline cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/tools-control.tsx:969]: If you mean that this code supports only `shape` and `track`, you should write exactly this.
Otherwise, when we will add one more object type, this part will be automatically broken and nobody will know about it

File: serverless/pytorch/omerferhatt/xmem/nuclio/function-gpu.yaml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                git \
+                ca-certificates \
+                python-is-python3 \
+                python3 \
+                python3-pip \
+              && rm -rf /var/lib/apt/lists/*
+        - kind: WORKDIR
+          value: /opt/nuclio
+        - kind: RUN
+          value: git clone --branch main https://github.com/omerferhatt/XMem xmem
```
Comment (bsekachev) [inline serverless/pytorch/omerferhatt/xmem/nuclio/function-gpu.yaml:40]: Why not to use original repository?

File: serverless/pytorch/omerferhatt/xmem/nuclio/function-gpu.yaml [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        - kind: RUN
+          value: pip install opencv-python-headless jsonpickle
+        - kind: RUN
+          value: |-
+            pip install torch torchvision
+        - kind: RUN
+          value: wget 'https://www.dropbox.com/scl/fi/5m1l747p15qzgq023e0q9/xmem.pth?rlkey=ss2kjaq4qlvvk5juucyvtmrh8&dl=0' -O '/xmem.pth'
```
Comment (bsekachev) [inline serverless/pytorch/omerferhatt/xmem/nuclio/function-gpu.yaml:47]: I do not think we should put the file to root `/`.
Better to put to current working directory.

File: serverless/pytorch/omerferhatt/xmem/nuclio/function-gpu.yaml [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+  triggers:
+    myHttpTrigger:
+      maxWorkers: 1
+      kind: 'http'
+      workerAvailabilityTimeoutMilliseconds: 10000
+      attributes:
+        # Set value from the calculation of tracking of 100 objects at the same time on a 4k image
+        maxRequestBodySize: 1073741824 # 1GB
```
Comment (bsekachev) [inline serverless/pytorch/omerferhatt/xmem/nuclio/function-gpu.yaml:56]: Is there a reason to send 1 Gb body, within HTTP request?

File: serverless/pytorch/omerferhatt/xmem/nuclio/function-gpu.yaml (hunk 4 above)
Comment (bsekachev) [reply serverless/pytorch/omerferhatt/xmem/nuclio/function-gpu.yaml:47]: I do not think dropbox is reliable enough.

Also it is not clear who is the weights owner, not clear weights license.
Is there original place?

File: serverless/pytorch/omerferhatt/xmem/nuclio/function-gpu.yaml [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+              && rm -rf /var/lib/apt/lists/*
+        - kind: WORKDIR
+          value: /opt/nuclio
+        - kind: RUN
+          value: git clone --branch main https://github.com/omerferhatt/XMem xmem
+        - kind: RUN
+          value: pip install opencv-python-headless jsonpickle
+        - kind: RUN
+          value: |-
+            pip install torch torchvision
```
Comment (bsekachev) [inline serverless/pytorch/omerferhatt/xmem/nuclio/function-gpu.yaml:45]: Usually it is a good idea to fix dependencies version. Otherwise sometimes it will turn around, that the image cannot be build anymore.

File: serverless/pytorch/omerferhatt/xmem/nuclio/function-gpu.yaml (hunk 3 above)
Comment (bsekachev) [reply serverless/pytorch/omerferhatt/xmem/nuclio/function-gpu.yaml:40]: Also it is better to use specific tag. `main` branch may be changed and new version will not compatible with old dependencies, for example.

**Comment by bsekachev** [comment]
Hello @omerferhatt 

Do you have any plans regarding the pull request?

**Comment by bsekachev** [comment]
I will close the pull request now.
Please, do not hesitate to reopen if you are ready to finish it.

**Comment by omerferhatt** [comment]
I'm going to try to finish this week, thanks.

**Comment by jtziebarth** [comment]
@omerferhatt @bsekachev 
I tried building CVAT with this feature and it works great overall.

However I notice that it takes a while to run inference on each frame (and in fact you can see it in the demo video omerferhatt posted above). Does anyone know what the source of this slow down is? When I run my own XMem inference script (adapted from the XMem author's [example](https://colab.research.google.com/drive/1RXK5QsUo2-CnOiy5AOSjoZggPVHOPh1m?usp=sharing) in their README), I can rip through inference on video at around 10 frames per second. With CVAT running on that same hardware, it takes several seconds to step through XMem tracking on a single frame. Is there something inherent in running XMem inference as a serverless function that causes the observed slowdown? And, if so, is there a way to speed it up? Otherwise, it will be very tedious to step through long videos frame by frame to track the objects. And this would be true of any model, not just XMem as used in this feature branch.

**Comment by omerferhatt** [comment]
> @omerferhatt @bsekachev I tried building CVAT with this feature and it works great overall.
> 
> However I notice that it takes a while to run inference on each frame (and in fact you can see it in the demo video omerferhatt posted above). Does anyone know what the source of this slow down is? When I run my own XMem inference script (adapted from the XMem author's [example](https://colab.research.google.com/drive/1RXK5QsUo2-CnOiy5AOSjoZggPVHOPh1m?usp=sharing) in their README), I can rip through inference on video at around 10 frames per second. With CVAT running on that same hardware, it takes several seconds to step through XMem tracking on a single frame. Is there something inherent in running XMem inference as a serverless function that causes the observed slowdown? And, if so, is there a way to speed it up? Otherwise, it will be very tedious to step through long videos frame by frame to track the objects. And this would be true of any model, not just XMem as used in this feature branch.

Hi @jtziebarth ,
Thanks for the recap. While I was working on it, I noticed that the reason it was running slow was the request time when importing and exporting states, but I didn't have much time to look into it. I'm currently working on SAM 2 because it's better than SAM + XMem++, so I don't think I can get back to this topic quickly. Sorry about that.

**Comment by jtziebarth** [comment]
@omerferhatt thanks for such a quick reply. Can you clarify what you mean by "request time when importing and exporting states?" Where in the code are you referring to? If you can point me in the right direction I can try to take a look at it. Thanks.
